### ISSUE TITLE
isfinite() in math.hpp causes build problems

### ISSUE THREAD

**Issue by davephillips** [VCVRack/Rack#318]
Linux Fedora 23, GCC 5.3.1
dev Rack

The following error occurs on any file with the algorithm include. Following a suggestion found at http://www.alecjacobson.com/weblog/?p=1768, commenting out the algorithm include allows the object files to build. However, another error occurs, reported in pt.2. 

...
In file included from include/rack.hpp:4:0,
                 from src/core/core.hpp:1,
                 from src/core/MidiCCToCV.cpp:4:
include/math.hpp: In member function ‘bool rack::Vec::isFinite()’:
include/math.hpp:207:20: error: ‘isfinite’ was not declared in this scope
   return isfinite(x) && isfinite(y);
                    ^
include/math.hpp:207:20: note: suggested alternative:
In file included from /usr/include/c++/5.3.1/random:38:0,
                 from /usr/include/c++/5.3.1/bits/stl_algo.h:66,
                 from /usr/include/c++/5.3.1/algorithm:62,
                 from src/core/MidiCCToCV.cpp:2:
/usr/include/c++/5.3.1/cmath:601:5: note:   ‘std::isfinite’
     isfinite(_Tp __x)
     ^
...
compile.mk:53: recipe for target 'build/src/core/MidiCCToCV.cpp.o' failed
make: *** [build/src/core/MidiCCToCV.cpp.o] Error 1

**Comment by bontric** [comment]
You are referencing a guide which starts with: 
> I’ve been trying to compile a C++ program with a set of header files one a Solaris SPARC machine. It’s my first experience with GCC outside of the usual intel/unix/mac world and it’s been a bumpy ride.

However, you are on Fedora, which **is** within the unix world.

Did you do a `make clean` before commenting out the algorithm? Did you try to do a "fresh" clone and make?

**Comment by davephillips** [comment]
Yes, I ran make clean and I have tried building from a fresh pull. Same problems.

**Comment by bontric** [comment]
Okay, so I looked a bit around and found that isfinite seems to be  somewhat problamatic (e.g. [here](https://bugs.freedesktop.org/show_bug.cgi?id=94761) or [here](https://github.com/open-source-parsers/jsoncpp/issues/214)) .  One suggestion I found was to replace `isfinite` with  `std::isfinite`.

Would be great if you could try this, so replace `math.hpp:207` with
```c
return std::isfinite(x) && std::isfinite(y);
```

**Comment by bontric** [comment]
Also: Could you please close the other issue and change the name of this issue to 
'isfinite() in math.hpp causes build problems`  or somethings similar? 

This is a bit more specific problem, glad you encountered it ;)

(Posting your `gcc` version might also help: `gcc --version`)

**Comment by davephillips** [comment]
Issue renamed. GCC is 5.3.1, it's in the first line of my first post. 

The problem remains:

g++ -Iinclude -Idep/include -Idep/lib/libzip/include -MMD -g -O3 -march=nocona -ffast-math -fno-finite-math-only -Wall -Wextra -Wno-unused-parameter -DARCH_LIN -Wsuggest-override -std=c++11 -c -o build/src/settings.cpp.o src/settings.cpp
In file included from include/widgets.hpp:9:0,
                 from include/app.hpp:4,
                 from src/settings.cpp:2:
include/math.hpp: In member function ‘bool rack::Vec::isFinite()’:
include/math.hpp:207:14: error: expected unqualified-id before ‘(’ token
   return std::isfinite(x) && std::isfinite(y);
              ^
In file included from include/widgets.hpp:9:0,
                 from include/app.hpp:4,
                 from src/settings.cpp:2:
include/math.hpp:207:34: error: expected unqualified-id before ‘(’ token
   return std::isfinite(x) && std::isfinite(y);
                                  ^
compile.mk:53: recipe for target 'build/src/settings.cpp.o' failed
make: *** [build/src/settings.cpp.o] Error 1

**Comment by DrChrisBoucher** [comment]
This works, tested on Linux Mint 18.2 and Windows:

```
diff --git a/include/math.hpp b/include/math.hpp
index 9f9f03e..bf372dd 100644
--- a/include/math.hpp
+++ b/include/math.hpp
@@ -2,7 +2,7 @@

 #include <stdint.h>
 #include <stdlib.h>
-#include <math.h>
+#include <cmath>


 namespace rack {
@@ -204,7 +204,7 @@ struct Vec {
                return x == 0.0 && y == 0.0;
        }
        bool isFinite() {
-               return isfinite(x) && isfinite(y);
+               return std::isfinite(x) && std::isfinite(y);
        }
        Vec clamp(Rect bound);
 };
```

**Comment by bontric** [comment]
@davephillips 
> GCC is 5.3.1, it's in the first line of my first post.

Sorry I overlooked that. Seems like it is definitely related to [this](https://bugs.freedesktop.org/show_bug.cgi?id=94761) then.

```bash
 include/math.hpp:207:34: error: expected unqualified-id before ‘(’ token
```
I might be wrong but that looks like a syntax error. Did you make any  other changes to `math.hpp`?

@DrChrisBoucher Thanks, is the switch to `cmath` include necessary though? It's a bit difficult to reproduce these problems for me, as I'm using `gcc 7.2.0` and any combination works for me..

**Comment by davephillips** [comment]
Thanks, Dr. Chris Boucher, it works for Fedora 23, GCC 5.3.1 as well. :) The build completes without problems now.

**Comment by DrChrisBoucher** [comment]
@bontric std::infinite is defined in cmath, so it's required. I've also tried building with math.h before and after cmath, and it worked in both cases, but since it builds with just cmath I removed it for clarity.

**Comment by bontric** [comment]
@DrChrisBoucher Thanks. Apparently it's also [preferable](https://stackoverflow.com/questions/15656434/include-cmath-vs-include-math-h-in-a-c-program) to use the c.. headers. 

Pinging @AndrewBelt, I guess it's not worth creating a PR for this

**Comment by AndrewBelt** [comment]
Fixed in abd194b. Technically no functions are defined by standard C++ functions at root level, but I find it annoying to prefix everything by `std::`, so I write standard non-compliant C++.
